### DISCUSSION TITLE
CSS selector where() causes issues with browsers

### DISCUSSION THREAD

**Discussion by ReduxFlakes** [withastro/roadmap#371] [Proposal]
As a person that uses a lot of browsers mainly Firefox and Gecko based ones I've noticed that some if not all Gecko based browser have issue with websites built on Astro specially loading CSS of, for example components and only loading external CSS files like global.css.

This is caused by these browsers not supporting the selector `:where()`. I'm not sure if we can replace it but at least try to find a replacement.

Down below is a comparison of the same website built on Astro in Firefox (LibreWold) and Pale Moon (Gecko based)

Thanks!

![Firefox](https://user-images.githubusercontent.com/87225951/199129732-c49c7c26-5bb6-4235-a1ab-d156b1d656d5.png)

![Pale Moon](https://user-images.githubusercontent.com/87225951/199129733-8f2e6cb8-2a46-42bd-b24b-bd504ca79058.png)

**Comment by phylor** [comment]
Is there are solution or ideas for this? Defining everything as `<style is:global>` works, but we then lose the ability of component focused CSS.

Would it be possible to have a configuration option not to use the `where` selector and just append the classes directly to the elements?

**Comment by andersk** [comment]
One workaround is to transform the `:where()` selectors using PostCSS. Note that this transformation alters the specificity of the selectors (which defeats Astro’s [motivation](https://github.com/withastro/rfcs/blob/main/proposals/0012-scoped-css-with-preserved-specificity.md) for using `:where()` in the first place).

```js
// postcss.config.cjs
exports.plugins = [
  {
    postcssPlugin: "inaccurate-pseudo-where",
    Rule(rule) {
      rule.selector = rule.selector?.replace(/:where\((\.astro-\w+)\)/g, "$1");
    },
  },
];
```

For reference, [browser support for `:where()`](https://caniuse.com/mdn-css_selectors_where) is about 92.75% as of December 2022: Chrome ≥ 88, Safari ≥ 14, Firefox ≥ 78, Opera ≥ 74. (Somewhat misleadingly, the [compiler PR](https://github.com/withastro/compiler/pull/330) featured a filtered view of this table.)

**Comment by bluwy** [comment]
Hello 👋 We have moved this proposal to stage 2! https://github.com/withastro/roadmap/issues/540

We will look into the better browser support for CSS scoping soon. Thanks for the discussion and feedback.
